### ISSUE TITLE
Sokol binaries for wasm

### ISSUE THREAD

**Issue by thePHTest** [floooh/sokol-odin#3]
I am starting to look into adding a build_clibs_wasm.cmd/sh script and get at least the microui example program working for web.
Are there any significant barriers to achieve this? It's my first foray into web assembly so it might take me a bit. Odin's wasm usage isn't as battle tested as other platforms at the moment either.

**Comment by floooh** [comment]
The sokol headers are using Emscripten specific functions and features (for instance the EM_JS macro to embed Javascript functions into C source code).

In Zig I was able to compile by using a combination of the Zig compiler and the Emscripten SDK (which is needed to provide a 'sysroot', and the linker), see here:

https://github.com/floooh/pacman.zig#experimental-web-support

...last time I checked this didn't work anymore though because of some breaking changes in the Emscripten SDK though, it may be that Zig needs to update to the latest LLVM version before it works again, not sure.

**Comment by thePHTest** [comment]
Ah interesting. Odin has the build flag `-target:js_wasm32` to compile for wasm. I believe there is some extra work that will need to be done for the Odin program to setup the allocators with wasm as the default is the nil allocator for that build target. I'll have to see if I need to do anything in particular to expose these `EM_JS` functions via Odin as well.

If there is anywhere you can point me to to grab an emscripten build command for sokol libs that would be great. Otherwise, I'll see what I can piece together.

**Comment by floooh** [comment]
Quick and dirty ```build_clibs_wasm.sh``` (see https://github.com/floooh/sokol-odin/blob/main/build_clibs_linux.sh for reference```):

```bash
set -e

build_lib_wasm_release() {
    src=$1
    dst=$2
    backend=$3
    echo $dst
    emcc -c -O2 -DNDEBUG -DIMPL -D$backend c/$src.c
    emar rcs $dst.a $src.o
}

build_lib_wasm_debug() {
    src=$1
    dst=$2
    backend=$3
    echo $dst
    emcc -c -g -DIMPL -D$backend c/$src.c
    emar rcs $dst.a $src.o
}

# wasm + GL + Release
build_lib_wasm_release sokol_gfx sokol/gfx/sokol_gfx_wasm_gl_release SOKOL_GLES3

# wasm + GL + Debug
build_lib_wasm_debug sokol_gfx sokol/gfx/sokol_gfx_wasm_gl_debug SOKOL_GLES3

rm *.o
```

Expects a 'global' install of the Emscripten SDK so that emcc and emar are in the path. I only tested compilation, not if the resulting link libraries actually work :)

**Comment by floooh** [comment]
PS: the tricky part will be the linking though... but let me know how it goes :)

**Comment by tyzor** [comment]
What options are there for removing emscripten requirements from the sokol bindings?  It seems like the EM_JS macros are the major hurdle and I see there was another discussion about it (https://github.com/floooh/sokol/issues/617)

**Comment by floooh** [comment]
> What options are there for removing emscripten requirements from the sokol bindings?

Yeah, that's not realistic option, it's not just the EM_JS magic, but all the other things provided by the Emscripten SDK and especially the Emscripten link step (like automatically creating a JS file for the 'web API glue' in the link step and running additional optimization via wasm-opt/binaryen and Closure). Doing all this stuff without Emscripten basically comes down to emulating Emscripten or switching to an alternative toolchain like https://github.com/schellingb/wajic (which is similar but much more bare bones).

**Comment by blob1807** [comment]
Should this issue be closed given there's working build scripts & a known way to get Odin + Sokol + Web working?
https://github.com/karl-zylinski/odin-sokol-web

**Comment by floooh** [comment]
Ah thanks for the hint, yeah I'll close the ticket :)
